modules/core/src/main/java/org/apache/ignite/internal/lock/AutoLockup.java

```diff
@@ -0,0 +1,33 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.lock;
+
+/**
+ * Represents a lockup (this is an aquisition and owning of a lock like {@link java.util.concurrent.locks.Lock})
+ * that plays nicely with try-with-resources clause. The underlying lock will be acquired on creation (or obtaining)
+ * a lockup, and the lock will be released when {@link AutoLockup#close()} method is called.
+ */
+@FunctionalInterface
+public interface AutoLockup extends AutoCloseable {
```

Comment:
This design looks like sugar, I don't think it's necessary.

Reply:
I think it's about habits. Everyone knows that try-with-resources is the way to go when some kind of a cleanup is required with any outcome, EXCEPTO FOR working with locks: everyone is used to the traditional `acquire(); try ... finally release()` idiom (probably because in many cases a try-with-resources would require an allocation). But in this case, it does not require an allocation, the standard idiom costs nothing, so why not use the more general idiom supported by the language? It also allows ot have one method less in the API (we don't need an explicit `release()` method).

Let's try it this way and see how it goes. If this becomes a problem, we can easily change it to the traditional approach with try/finally.